### PR TITLE
MixupCallback

### PR DESCRIPTION
**Pull request by lightforever**
The callback for the mixup.

fields: features which must be affected by the mixup.

flpha: Beta distribution a=b parameters

train_only: boolean. Apply to train only? As the mixup use the proxy inputs, the targets are also proxy.
We are not interested in it, are we? So, if train_only is False, use a standard output/metric.

### CODE REVIEW

File: catalyst/dl/callbacks/mixup.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from catalyst.dl.callbacks import CriterionCallback
+import numpy as np
+import torch
+
+
+class MixupCallback(CriterionCallback):
+    def __init__(self, fields: List[str] = ('features',), alpha=1.0, train_only=True, **kwargs):
+        assert alpha >= 0, 'alpha must be>=0'
+
+        super(MixupCallback, self).__init__(**kwargs)
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: for py3 you can use just `super().__init__` to follow lib code standarts

File: catalyst/dl/callbacks/mixup.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        self.train_only = train_only
+        self.fields = fields
+        self.alpha = alpha
+        self.lam = 1
+        self.index = None
+
+    def on_batch_start(self, state: RunnerState):
+        if self.train_only and state.loader_name != 'train':
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: it better to use `state.loader_name.startswith('train')`

PS. btw, please use " to follow lib code standarts

File: catalyst/dl/callbacks/mixup.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        self.index = torch.randperm(state.input[self.fields[0]].shape[0])
+        if state.input[self.fields[0]].is_cuda:
+            self.index = self.index.cuda()
+
+        for f in self.fields:
+            state.input[f] = self.lam * state.input[f] + (1 - self.lam) * state.input[f][self.index]
+
+    def _compute_loss(self, state: RunnerState, criterion):
+        if self.train_only and state.loader_name != 'train':
+            return super(MixupCallback, self)._compute_loss(state, criterion)
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: can be simplified to `super().` (py3)

File: catalyst/dl/callbacks/mixup.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    def on_batch_start(self, state: RunnerState):
+        if self.train_only and state.loader_name != 'train':
+            return
+        if self.alpha > 0:
+            self.lam = np.random.beta(self.alpha, self.alpha)
+        else:
+            self.lam = 1
+
+        self.index = torch.randperm(state.input[self.fields[0]].shape[0])
+        if state.input[self.fields[0]].is_cuda:
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: what do you think about such simplification:`self.index = self.index.to(state.device)`?

File: catalyst/dl/callbacks/mixup.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            state.input[f] = self.lam * state.input[f] + (1 - self.lam) * state.input[f][self.index]
+
+    def _compute_loss(self, state: RunnerState, criterion):
+        if self.train_only and state.loader_name != 'train':
+            return super(MixupCallback, self)._compute_loss(state, criterion)
+        pred = state.output[self.output_key]
+        y_a = state.input[self.input_key]
+        y_b = state.input[self.input_key][self.index]
+
+        return self.lam * criterion(pred, y_a) + (1 - self.lam) * criterion(pred, y_b)
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: to allow easy debug for future needs it's better to make something like
```
loss = ...
return loss
```

File: catalyst/dl/callbacks/mixup.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+from typing import List
+from catalyst.dl.core.state import RunnerState
+from catalyst.dl.callbacks import CriterionCallback
+import numpy as np
+import torch
+
+
+class MixupCallback(CriterionCallback):
+    def __init__(self, fields: List[str] = ('features',), alpha=1.0, train_only=True, **kwargs):
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: please, update the docs for new callback

File: catalyst/dl/callbacks/mixup.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+from typing import List
+from catalyst.dl.core.state import RunnerState
+from catalyst.dl.callbacks import CriterionCallback
+import numpy as np
+import torch
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py:5]: to follow lib code standarts, its better to use the following import sequence:
- core python libs (typing, collection, etc)
- python libs (numpy, etc)
- dl libs (pytorch, etc)
- catalyst imports

**Review by Scitator — CHANGES_REQUESTED** [review]

File: catalyst/dl/callbacks/mixup.py (hunk 1 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: That is great information, thanks. Corrected

File: catalyst/dl/callbacks/mixup.py (hunk 2 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: Fixed. What is the reason to use double quotes instead of a single one? 
To type " we should use Shift on the most keyboards. Do you have one that allows doing the type in a single press ?)

File: catalyst/dl/callbacks/mixup.py (hunk 3 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: That is great information, thanks. Corrected

File: catalyst/dl/callbacks/mixup.py (hunk 4 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: This is much better. Corrected

File: catalyst/dl/callbacks/mixup.py (hunk 5 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: Fixed

File: catalyst/dl/callbacks/mixup.py (hunk 6 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: Done

File: catalyst/dl/callbacks/mixup.py (hunk 7 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py:5]: Done, thanks.

File: catalyst/dl/callbacks/mixup.py [hunk 8]
```diff
@@ -1,50 +1,56 @@
...
+        super().__init__(**kwargs)
 
         self.train_only = train_only
         self.fields = fields
         self.alpha = alpha
         self.lam = 1
         self.index = None
 
+    def _is_needed(self, state: RunnerState):
+        return not self.train_only or state.loader_name.startswith("train")
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: what about check this flag `on_loader_start`?

File: catalyst/dl/callbacks/mixup.py [hunk 9]
```diff
@@ -1,50 +1,56 @@
...
+
     def on_batch_start(self, state: RunnerState):
-        if self.train_only and state.loader_name != 'train':
+        if not self._is_needed(state):
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: what about check this flag `on_loader_start`?

File: catalyst/dl/callbacks/mixup.py [hunk 10]
```diff
@@ -1,50 +1,56 @@
...
+
 import torch
 
+from catalyst.dl.core.state import RunnerState
+from catalyst.dl.callbacks import CriterionCallback
+
 
 class MixupCallback(CriterionCallback):
-    def __init__(self, fields: List[str] = ('features',), alpha=1.0,
+    def __init__(self, fields: List[str] = ("features",), alpha=1.0,
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: add docs in [Google format](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md#documentation) please

File: catalyst/dl/callbacks/mixup.py (hunk 8 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: Thanks, done.

File: catalyst/dl/callbacks/mixup.py (hunk 9 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: Thanks, done.

File: catalyst/dl/callbacks/mixup.py (hunk 10 above)
Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: Thanks, done.

File: catalyst/dl/callbacks/mixup.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    Paper: https://arxiv.org/abs/1710.09412
+
+    Note:
+        MixupCallback is inherited from CriterionCallback and
+        does its work.
+
+        You may not use them together.
+    """
+    def __init__(self, fields: List[str] = ("features",), alpha=1.0,
+                 train_only=True, **kwargs):
```
Comment (Scitator) [inline catalyst/dl/callbacks/mixup.py]: could you please refactor to 
```
def __init__(
    self, 
    fields: List[str] = ("features",), 
    alpha=1.0,
    train_only=True, 
    **kwargs
):
```
one more common lib standard :)

Comment (lightforever) [reply catalyst/dl/callbacks/mixup.py]: Done :)

**Comment by Scitator** [comment]
It looks great!
As a final change, could you please add config3.yml - a copy of config1.yml with `MixupCallback` (rather than `CriterionCallback`) to travis CI?

All you need is to add
```
- PYTHONPATH=./examples:./catalyst:${PYTHONPATH}
          python catalyst/dl/scripts/run.py
          --expdir=./examples/_tests_mnist_stages2
          --config=./examples/_tests_mnist_stages2/config3.yml
          --logdir=./examples/logs/_tests_mnist_stages2
          --check
```
after https://github.com/catalyst-team/catalyst/blob/master/.travis.yml#L94

and contribute config3.yml with your `MixupCallback`

PS. config1 could be found [here](https://github.com/catalyst-team/catalyst/blob/master/examples/_tests_mnist_stages2/config1.yml)

**Comment by lightforever** [comment]
@Scitator, Done :)

**Review by Scitator — APPROVED** [review]
